File: dev/requirements/shared/Saga.swift
#if canImport(SwiftUI) && os(iOS)
import SwiftUI
import UIKit
import SetlistCore
import SetlistUI

/// One beat of a story: what the comedian does, and the caption that narrates
/// it. The caption is written in user terms, because it is what surfaces in the
/// spec's gallery beneath the frame it labels.
struct SagaStep {
    let caption: String
    let act: (inout StageState) -> Void

    init(_ caption: String, _ act: @escaping (inout StageState) -> Void) {
        self.caption = caption
        self.act = act
    }
}

/// A saga proves what a single frame cannot: what arriving, tapping or time
/// passing *changes*. Each step drives the same real screen every other kind
/// renders, and one frame is captured after each — never a slideshow of
/// hand-arranged states.
func sagaCase(
    id: String,
    slug: String,
    from initial: @escaping () -> StageState,
    steps: [SagaStep]
) -> RequirementCase {
    RequirementCase(id: id) {
        try MainActor.assumeIsolated {
            var state = initial()
            var captions: [String] = []
            for (index, step) in steps.enumerated() {
                step.act(&state)
                try expectScreen(
                    StageScreen(state: state),
                    slug: slug,
                    id: id,
                    frame: index + 1
                )
                captions.append(step.caption)
            }
            try writeCaptions(captions, slug: slug, id: id)
        }
    }
}

/// The captions ride beside the frames as generated data: the gallery needs
/// them to label the strip, and Markdown is not a place to hand-maintain a
/// list that must stay in step with the frames.
private func writeCaptions(_ captions: [String], slug: String, id: String) throws {
    let url = REPO_ROOT.appendingPathComponent("dev/requirements/saga/cases/\(slug).\(id).captions.json")
    let data = try JSONSerialization.data(withJSONObject: captions, options: [.prettyPrinted, .withoutEscapingSlashes])
    let rendered = String(decoding: data, as: UTF8.self) + "\n"

    guard !isRefreshingGoldens else {
        try rendered.write(to: url, atomically: true, encoding: .utf8)
        return
    }
    let committed = try? String(contentsOf: url, encoding: .utf8)
    guard committed == rendered else {
        throw RequirementFailure(
            description: "\(slug).\(id): the captions beside the frames are out of date — refresh"
        )
    }
}
#endif

File: dev/requirements/shared/saga.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:setlist_core/setlist_core.dart';
import 'package:setlist_ui/setlist_ui.dart';
import 'package:setlist_ui/testing.dart';

import 'requirement_case.dart';

/// One beat of a story: what happens, the caption that narrates it, and the
/// part of the screen it changes.
///
/// The caption is written in user terms, because it is what surfaces in the
/// spec's gallery beneath the frame it labels. The region is what the frame is
/// cropped to: a storyboard of six whole screens costs a reader the section and
/// shows them mostly what did not move. A step that genuinely changes the whole
/// screen leaves it null and says so by doing.
class SagaStep {
  const SagaStep(this.caption, {this.region, required this.act});

  final String caption;
  final StageRegion? region;

  /// Drives the model forward and answers where the set clock now stands.
  final Duration Function(StageModel model, Duration elapsed) act;
}

/// A saga proves what a single frame cannot: what arriving, tapping or time
/// passing *changes*.
///
/// Each step drives the same real screen every other kind renders, and one
/// frame is captured after each — never a slideshow of hand-arranged states.
RequirementCase sagaCase({
  required String id,
  required String slug,
  required StageModel Function() from,
  required List<SagaStep> steps,
}) {
  return RequirementCase(id, (tester) async {
    final model = from();
    var elapsed = Duration.zero;
    final captions = <String>[];

    for (var index = 0; index < steps.length; index++) {
      final step = steps[index];
      elapsed = step.act(model, elapsed);
      final state = StageState(
        model: model,
        venue: StageFixture.venue,
        elapsed: elapsed,
      );

      await pumpStage(tester, state);
      final frame = (index + 1).toString().padLeft(2, '0');
      await expectLater(
        step.region == null
            ? find.byType(StageScreen)
            : find.byKey(step.region!.key),
        matchesGoldenFile('saga/cases/$slug.$id.step-$frame.png'),
      );
      captions.add(step.caption);
    }

    // Generated beside the frames rather than hand-maintained in Markdown: a
    // caption list written by hand drifts from the frames it labels, and the
    // gate fails when their counts disagree.
    File('saga/cases/$slug.$id.captions.json')
        .writeAsStringSync('${const JsonEncoder.withIndent('  ').convert(captions)}\n');
  });
}
